include/revng/AutoEnforcer/AutoEnforcer.h

```diff
@@ -0,0 +1,183 @@
+//
+
+#include <string>
+#include <utility>
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/AutoEnforcer/AutoEnforcerContext.h"
+#include "revng/AutoEnforcer/AutoEnforcerTarget.h"
+#include "revng/AutoEnforcer/BackingContainerRegistry.h"
+#include "revng/AutoEnforcer/InvalidationEvent.h"
+#include "revng/AutoEnforcer/KindsRegistry.h"
+#include "revng/AutoEnforcer/Pipeline.h"
+#include "revng/Support/Debug.h"
+
+namespace AutoEnforcer {
+
+class PipelineRunner {
+public:
+  using iterator = std::vector<Step>::iterator;
+  using const_iterator = std::vector<Step>::const_iterator;
+  using InvalidationMap = llvm::StringMap<BackingContainersStatus>;
+
+  explicit PipelineRunner(Context &C) : AEContext(&C) {}
+
+  template<typename... EnforcerWrappers>
+  void addStep(std::string StepName, EnforcerWrappers &&...EnfWrappers) {
+    CommittedRegistry = true;
+    Pipeline.add(Step(std::move(StepName),
+                      Registry.createEmpty(),
+                      std::forward<EnforcerWrappers>(EnfWrappers)...));
+  }
+
+  llvm::Error deduceInvalidations(InvalidationMap &Invalidations) const {
+    return Pipeline.getInvalidations(Invalidations);
+  }
+
+  const BackingContainers &getStartingContainers() const {
+    return Pipeline.getStartingContainers();
+  }
+
+  BackingContainers &getStartingContainers() {
+    return Pipeline.getStartingContainers();
+  }
+
+  template<typename BackingContainerType>
+  const BackingContainerType &getStartingContainer(llvm::StringRef Name) const {
+    return Pipeline.getStartingContainer<BackingContainerType>(Name);
+  }
+
+  template<typename BackingContainerType>
+  BackingContainerType &getOrCreateStartingContainer(llvm::StringRef Name) {
+    return Pipeline.getOrCreateStartingContainer<BackingContainerType>(Name);
+  }
+
+  bool isStartingContainerEmpty(llvm::StringRef Name) const {
+    return Pipeline.getStartingContainers().isContainerEmpty(Name);
+  }
+
+  const BackingContainers &getFinalContainers() const {
+    return Pipeline.getFinalContainers();
+  }
+  BackingContainers &getFinalContainers() {
+    return Pipeline.getFinalContainers();
+  }
+
+  template<typename BackingContainerType>
+  const BackingContainerType &getFinalContainer(llvm::StringRef Name) const {
+    return Pipeline.getFinalContainer<BackingContainerType>(Name);
+  }
+
+  template<typename BackingContainerType>
+  BackingContainerType &getOrCreateFinalContainer(llvm::StringRef Name) {
+    return Pipeline.getOrCreateFinalContainer<BackingContainerType>(Name);
+  }
+
+  llvm::Error
+  run(const BackingContainersStatus &Targets, llvm::raw_ostream *OS = nullptr) {
+    return Pipeline.run(*AEContext, Targets, OS);
+  }
+
+  template<typename BackingContainerType, typename... Args>
+  void registerContainerFactory(llvm::StringRef Name, Args &&...Arguments) {
```

Comment:
This sounds more like *creation* than *registration*. What do you think?
```suggestion
  void createContainerFactory(llvm::StringRef Name, Args &&...Arguments) {
```